### ISSUE TITLE
Class Name is lost when creating new one by duplicating existing script

### ISSUE THREAD

**Issue by NaleVex** [godotengine/godot#83395]
### Godot version

4.1.2

### System information

Win 10

### Issue description

After creating new script by duplicating script with already made custom class_name, that class_name is lost inside godot.

### Steps to reproduce

Exists a script with any custom created resource class_name
for example we have this script
```
extends Resource
class_name NewResource
```
After creating this script we now can use NewResrource as resource, it's registered inside Godot.
Now if we want to create another class_name by duplicating existing one, new script will have same header with
```
extends Resource
class_name NewResource
```
and even if we rename `class_name` to something else in new script, class_name from original script is lost inside godot.
To register back old class_name we should go back into that script, make some changes to class_name name and save the script again.

### Minimal reproduction project

"N/A"

**Comment by AThousandShips** [comment]
This is kind of unavoidable, only one script can have a specific `class_name`, you will need to copy it manually instead

**Comment by NaleVex** [comment]
> This is kind of unavoidable, only one script can have a specific `class_name`, you will need to copy it manually instead

Indeed, but since new script has `class_name` changed, `class_name` for old script should be working back again.
I'd suggest, if something ever happens when in new script there are `class_name` that has the name that is **already** in some other older scripts, that name in new script errored and ignored by godot leaving keeping old one working, not removing old one and accepting new instead.

**Comment by AThousandShips** [comment]
That would be highly complex and hard to do, see #76380

**Comment by NaleVex** [comment]
Highly comples is what I just suggested? If whenever new class_name created and that name is already registered, do not `unregister` that one and error this one instead? It shouldn't require to reload every script to check their name, only that one which is already registered with that same name.

**Comment by lostminds** [comment]
I've also run into this a number of times, and it's annoying since duplicating an existing class script feels like a natural way to start making a new similar class. And it's initially surprising that this breaks (at least temporarily) the old class, even if I think I now understand why it happens.

Based on the basic explanation of the system here: https://github.com/godotengine/godot/issues/76380#issuecomment-1574264774 perhaps a way to handle this would be to allow the system that tracks class_name-script mapping to store multiple scripts per class_name. When compiling this list will be checked, and for any class_names with multiple scripts assigned to it will check these scripts to see if the scripts are still actually referencing the class_name and if the referenced scripts even exist (for move cases). If not, the outdated mapping are removed. If there are still multiple scripts mapped to a single class_name, an error is thrown explaining that multiple scripts (and what scripts) are referencing this class_name and that that's not allowed.

I think this could solve the issue both when duplicating scripts as described above and when moving scripts like in https://github.com/godotengine/godot/issues/76380 and as a bonus we could also get a better error message explaining what we've done wrong and how to fix it when there are real duplicate class_name references.

**Comment by AThousandShips** [comment]
But that would require you to track old classes with the same name, how would you know which one is valid?

**Comment by AThousandShips** [comment]
IMO the solution is a warning when duplicating a script like this, including when copying it in the file system, allowing you to not import it and fix it

**Comment by AThousandShips** [comment]
As I said I don't think that's a good idea because it makes for a lot of complexity where none is needed IMO

**Comment by RobRuana** [comment]
The behavior in 3 was to keep the mapping to the original script file, and only add a mapping for the duplicated script when class_name is updated. Seems like the old behavior was both simple and what the user would expect: if a mapping already exists to a valid script file, then don't update the mapping to the duplicate script file.

btw, when this happens to me, I solve it by manually editing `.godot/global_script_class_cache.cfg` after I've updated class_name in the duplicated file.

**Comment by daenvil** [comment]
I just had this bug crash my whole project again. Duplicating an existing class and renaming it seems like a totally normal workflow and it's weird that Godot gives problems with this. IMO ignoring the issue or just throwing a warning to the issue are not appropiate solutions. The logical thing would be to the new duplicated class to throw an error, instead of the old one which was already registered.

**Comment by HolonProduction** [comment]
Can't seem to reproduce in v4.7.1.stable.flathub [a13da4feb]

("that class_name is lost inside godot" it's hard to know what to test here, I tried with the create dialog for an exported `Resource` property)
